[WIP] Airy's Equation and Bessel's equation solving

nagarakshay opened the pull request:
Implementation of Airy's equation and Bessel's equation

Fixes #9819 
[1]Airy's equation

```
Eq(x ** 2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second kind respectively
```

[2]Bessel's equation

```
Eq((x+a)**2 * Derivative(y(x),x,x)+(x+a) * Derivative(y(x),x)+(λ**2* (x-a)**2-n ** 2)*y(x))
The above equation now has solution c1* besselj(n,λ*x - λ*a)+c2 *bessely(n,λ*x - λ*a)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second kind respectively
```

Commit:
Implementation of Airy's equation and Bessel's equation

Fixes #9819
[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

Commit:
Merge branch 'master' of http://github.com/sympy/sympy

Due to changes in sympy

Commit:
Implementation of Airy's equation and Bessel's equation

[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

nagarakshay commented:
@zanzibar7 @cbm755 Ping!

Commit:
Implementation of Airy's equation and Bessel's equation

[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

zanzibar7 commented on sympy/solvers/ode.py:
```diff
@@ -303,7 +303,9 @@
     "Liouville",
+    "2nd_linear_airy",
     "2nd_power_series_ordinary",
+    "Bessel_Equation",
```
Better to name new hints consistently with existing names.  I suggest "2nd_linear_bessel"

zanzibar7 commented on sympy/solvers/ode.py:
```diff
@@ -3780,6 +3825,17 @@ def ode_2nd_power_series_regular(eq, func, order, match):
             return Eq(f(x), collect(finalseries1 + finalseries2,
                 [C0, C1]) + Order(x**terms))
 
+def ode_Bessel_Equation(eq, func, order, match):
```
Suggest `ode_2nd_linear_bessel` .

zanzibar7 commented:
Looks like the tests failed.  Before submitting a pull request, run the test suite locally and make sure things pass at your end with `bin/test` and `bin/doctest`.

We also need to enhance `classify_ode` under `order == 2` to recognize Airy and Bessel equations, so that these get called correctly.

And we need tests in `test_ode.py` to show 

* Equations are appropriately classified
* Equations are solved correctly

Commit:
Implementation of Airy's equation and Bessel's equation

Fixes #9819
[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

Also test cases and docstrings have been added.

Commit:
Implementation of Airy's equation and Bessel's equation

[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

Tests and docstrings have been added.
Also the function has been renamed to 2nd_linear_bessel

nagarakshay commented:
@zanzibar7 Ok now the tests pass from my side. I have amended the name of equation too and added the tests too. Is there anything else?

zanzibar7 commented on sympy/solvers/ode.py:
```diff
@@ -1202,8 +1203,25 @@ class in it.  Note that a hint may do this anyway if
+                            #If the ODE has regular singular point at x0 and is of the form
+                            #Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x) thus Bessel's equation
+                            if p==1:
+                                b4 = Wild('b4', exclude=[x,f(x),df])
```
Does b4 have to be an integer, or will this work with fractional Bessel orders?  

Also, can we catch equations that are not already in standard form?

nagarakshay replied:
Yes b4 can be anything. No, please elaborate what sort of equations should the Bessel's equation support?

Commit:
Implementation of Airy's equation and Bessel's equation

[1]Airy's equation
Eq(x**2*Derivative(y(x),x,x)-(ax+b)*y(x))
The above equation now has solution c1*Ai(x)+x2*Bi(x)
where Ai(x) and Bi(x) are Airy's functions of first and second order respectively

[2]Bessel's equation
Eq(x**2*Derivative(y(x),x,x)+x*Derivative(y(x),x)+(x**2-n**2)*y(x))
The above equation now has solution c1*besselj(n,x)+c2*bessely(n,x)
where besselj(n,x) and bessely(n,x) are Bessel's functions of first and second order respectively

Tests and docstrings have been added and revised ( keeping in consideration the different ways integers and non integer n's are treated in Bessel's equation )
Also support for n=0 has been added

jksuom commented:
> what sort of equations should the Bessel's equation method of solving support?

It seems that the class of equations whose solutions can be expressed in terms of Bessel functions is large. To take them all would be quite a big task. But I think that at least those equations could be included that have solutions of the type ``J(2*x)`` or ``J(x + 1)`` where ``J`` is a Bessel function. That is, equations obtained from Bessel's equation by a simple change of variable.

nagarakshay commented:
Sure, I will add it right away.

Commit:
Changing the Bessel's function to check for change of parameters

Specifically changing parameter a4*x to x and then after solving substituting it back
>>>eq = t**2*(f(t).diff(t, 2)) + t*(f(t).diff(t)) + 25*t**2*f(t)
>>>C0, C1, C2 = get_numbered_constants(eq, num=3)
>>>print(dsolve(eq))
Eq(f(5*t), C1*besselj(0, 5*t) + C2*bessely(0, 5*t))
>>>eq = t**2*(f(t).diff(t, 2)) + t*(f(t).diff(t)) + (25*t**2-36)*f(t)
>>>print(dsolve(eq))
Eq(f(5*t), C1*besselj(6, 5*t) + C2*bessely(6, 5*t))

nagarakshay commented:
@jksuom @zanzibar7 Does this look ok?

jksuom commented:
I think that one should start by looking for the singular points. If there is only one, then that could be used as ``x0``.

nagarakshay commented:
As you say, I started to implemented the idea but there was a small hindrance.
Here is what happened effectively.
 I passed the equation 
>>eq = (x-2)**2*f(x).diff(x, 2)+ (x-2)*(f(x).diff(x)) + 25*(x-2)**2*f(x)
>>a3 = Wild('a3', exclude=[f(x), df, f(x).diff(x, 2)])
    b3 = Wild('b3', exclude=[f(x), df, f(x).diff(x, 2)])
    c3 = Wild('c3', exclude=[f(x), df, f(x).diff(x, 2)])
>>reduced_exp
25*x**2*f(x) + x**2*Derivative(f(x), x, x) - 100*x*f(x) + x*Derivative(f(x), x) - 4*x*Derivative(f(x), x, x) + 100*f(x) - 2*Derivative(f(x), x) + 4*Derivative(f(x), x x)
>>deq = b3*df + a3*(f(x).diff(x, 2)) + c3*f(x)
>>r = collect(reduced_eq,[f(x).diff(x, 2), f(x).diff(x), f(x)]).match(deq)
None
I cant understand why is Null being returned?

nagarakshay commented:
Ping @jksuom

jksuom commented:
It seems to me that None was returned because ``reduced_eq`` is not of the type expected by ``deq``.

nagarakshay commented:
@jksuom I didn't get you. After collect(reduced_eq,[f(x).diff(x, 2), f(x).diff(x), f(x)]
 the expression should be of the form required by deq.

jksuom commented:
That was only a guess since it is hard to see what the actual values are. The code should be surrounded by triple backticks to preserve the original form.

nagarakshay commented:
I have a question. Do brackets matter while matching two expression? Or in other words if collect returns a parenthesizedpolynomial will a3 match it?

jksuom commented:
I do not know how match behaves. I suspect that match does not see brackets or parentheses. But you should experiment and investigate that.

jksuom commented:
It seems that match is not able to deal with more complicated expressions, because it does internally expand them even if the coefficients have been collected into a parenthesized expression. Probably you should try to find the coefficients by some other method. Perhaps by analysing the arguments of the collected equation in the code.

nagarakshay commented:
I matched all of the Derivatives individually and now it works perfectly also i amended the regular power series code to make it work with arbitary singular points (it previously had no mechanism for that)but some other tests are failing due to which this PR is getting delayed. I will try to address them as soon as possible.

nagarakshay commented:
@jksuom Is there a way of converting an object of class Derivative to an object of class Add?
Actually when the function has only a single Derivative
```
>>>reduced_exp=Derivative(f(x),x,2)
>>>reduced_exp.args #This should only return the current derivative
[f(x),x,x]

jksuom commented:
You could probably use ``Add.make_args(reduced_exp)``.

nagarakshay commented:
@jksuom Yeah that works perfectly. Thanks

asmeurer commented:
What is the status of this?

nagarakshay commented:
@asmeurer Sorry for the delay. I was busy with my university exams and end term assessments. I have almost completed this PR 2-3 test cases were failing. I will update in 2 days max.

RituRajSingh878 commented:
what's the status of this pr .can I start from the beginning by new pr.

oscarbenjamin commented:
I would think that it is worth starting from this PR if you want to work on the same thing.
